internal/lockfilescalibr/vcs/gitrepo/extractor.go

```diff
@@ -0,0 +1,120 @@
+package gitrepo
+
+import (
+	"context"
+	"io/fs"
+	"path"
+	"path/filepath"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/google/osv-scalibr/extractor"
+	"github.com/google/osv-scalibr/extractor/filesystem"
+	"github.com/google/osv-scalibr/plugin"
+	"github.com/google/osv-scalibr/purl"
+)
+
+type Extractor struct{}
+
+var _ filesystem.Extractor = Extractor{}
+
+func getCommitSHA(repo *git.Repository) (string, error) {
+	head, err := repo.Head()
+	if err != nil {
+		return "", err
+	}
+
+	return head.Hash().String(), nil
+}
+
+func getSubmodules(repo *git.Repository) (submodules []*git.SubmoduleStatus, err error) {
+	worktree, err := repo.Worktree()
+	if err != nil {
+		return nil, err
+	}
+	ss, err := worktree.Submodules()
+	if err != nil {
+		return nil, err
+	}
+	for _, s := range ss {
+		status, err := s.Status()
+		if err != nil {
+			continue
+		}
+		submodules = append(submodules, status)
+	}
+
+	return submodules, nil
+}
+
+func createCommitQueryInventory(commit string, path string) *extractor.Inventory {
+	return &extractor.Inventory{
+		SourceCode: &extractor.SourceCodeIdentifier{
+			Commit: commit,
+		},
+		Locations: []string{path},
+	}
+}
+
+// Name of the extractor.
+func (e Extractor) Name() string { return "vcs/gitrepo" }
+
+// Version of the extractor.
+func (e Extractor) Version() int { return 0 }
+
+// Requirements of the extractor.
+func (e Extractor) Requirements() *plugin.Capabilities {
+	return &plugin.Capabilities{}
+}
+
+// FileRequired returns true for .package-lock.json files under node_modules
+func (e Extractor) FileRequired(path string, fi fs.FileInfo) bool {
+	return fi.IsDir() && filepath.Base(path) == ".git"
+}
+
+// Extract extracts packages from yarn.lock files passed through the scan input.
+func (e Extractor) Extract(_ context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
+	// Assume this is fully on a real filesystem
+	// TODO: Make this support virtual filesystems
```

Comment:
Is there a lot of glue code that would have to be written to make this work?

Reply:
yeah, git also uses a virtual filesystem, but with different signatures compared to our filesystem from what I can tell. Need to look into this more.